scripts/base/protocols/redis/main.zeek

```diff
@@ -0,0 +1,258 @@
+		no_response_ranges: vector of NoReplyRange;
+		## We store if this analyzer had a violation to avoid logging if so.
+		## This should not be super necessary, but worth a shot.
+		violation: bool &default=F;
+	};
+
+	# Redis specifically mentions 10k commands as a good pipelining threshold, so
+	# we'll piggyback on that.
+	option max_pending_requests = 10000;
+}
+
+redef record connection += {
+	redis: Info &optional;
+	redis_state: State &optional;
+};
+
+redef likely_server_ports += {ports};
+
+event zeek_init() &priority=5
+	{
+	Log::create_stream(Redis::LOG, [$columns=Info, $path="redis",
+	    $policy=log_policy]);
+
+	Analyzer::register_for_ports(Analyzer::ANALYZER_SPICY_REDIS, ports);
+	}
+
+event analyzer_violation_info(atype: AllAnalyzers::Tag,
+    info: AnalyzerViolationInfo)
+	{
+	if ( atype == Analyzer::ANALYZER_SPICY_REDIS && info?$c && info$c
+	    ?$redis_state )
+		{
+		info$c$redis_state$violation = T;
+		}
+	}
+
+function new_redis_info(c: connection): Info
+	{
+	return Info($ts=network_time(), $uid=c$uid, $id=c$id);
+	}
+
+function make_new_state(c: connection)
+	{
+	local s: State;
+	c$redis_state = s;
+	Conn::register_removal_hook(c, finalize_redis);
+	}
+
+function set_state(c: connection, is_orig: bool)
+	{
+	if ( ! c?$redis_state )
+		make_new_state(c);
+
+	local current: count;
+	if ( is_orig )
+		current = c$redis_state$current_request;
+	else
+		current = c$redis_state$current_response;
+
+	if ( current !in c$redis_state$pending )
+		c$redis_state$pending[current] = new_redis_info(c);
+
+	c$redis = c$redis_state$pending[current];
+	}
+
+## Returns whether the last "no response" interval is not still open.
+function is_last_interval_closed(c: connection): bool
+	{
+	return |c$redis_state$no_response_ranges| == 0 ||
+	    c$redis_state$no_response_ranges[|c$redis_state$no_response_ranges| - 1]
+	    ?$end;
+	}
+
+event Redis::client_command(c: connection, command: Command)
+	{
+	if ( ! c?$redis_state )
+		make_new_state(c);
+
+	if ( max_pending_requests > 0
+	    && |c$redis_state$pending| > max_pending_requests )
+		{
+		Reporter::conn_weird("Redis_excessive_pipelining", c);
+		# Delete the current state and restart later. We'll be in a weird state, but
+		# really we want to abort. I don't quite get how to register this as a
+		# violation. :)
+		delete c$redis_state;
+		return;
+		}
+
+	++c$redis_state$current_request;
+	# CLIENT commands can skip a number of responses and may be used with
+	# pipelining. We need special logic in order to track the request/response
+	# pairs.
+	if ( command?$known && command$known == KnownCommand_CLIENT )
+		{
+		# All 3 CLIENT commands we care about have 3 elements
+		if ( |command$raw| == 3 )
+			{
+			if ( to_lower(command$raw[2]) == "on" )
+				{
+				# If the last range is open, close it here. Otherwise, noop
+				if ( |c$redis_state$no_response_ranges| > 0 )
+					{
+					local range = c$redis_state$no_response_ranges[|c$redis_state$no_response_ranges|
```

Comment:
```suggestion
					local range = c$redis_state$no_response_ranges[-1];
```

Thought this works today.

Reply:
Yeah, since Zeek 6. I just stumbled over that yesterday when trying to write code that works with 5.2! :-)